canvas_sdk/data/data_access_layer_client.py

```diff
@@ -0,0 +1,87 @@
+"""
+Data Access Layer client.
+
+This module is primarily responsible for executing calls to the gRPC service so that such details
+are abstracted away from callers. The return values of the methods on the client class are protobufs
+which must be mapped to user-facing objects.
+"""
+
+import functools
+from collections.abc import Callable
+from types import FunctionType
+from typing import Any
+
+import grpc
+from grpc import StatusCode
+
+from canvas_generated.data_access_layer.data_access_layer_pb2 import ID, Patient
+from canvas_generated.data_access_layer.data_access_layer_pb2_grpc import (
+    DataAccessLayerStub,
+)
+from settings import DAL_TARGET
+
+from . import exceptions
+
+
+class _DataAccessLayerClientMeta(type):
+    """
+    Metaclass for the Data Access Layer client class.
+
+    Wraps all methods of a class with a gRPC error handler.
+    """
+
+    def __new__(cls, name: str, bases: tuple, attrs: dict) -> type:
+        for name, value in attrs.items():
+            if isinstance(value, FunctionType):
+                attrs[name] = cls.handle_grpc_errors(value)
+        return super().__new__(cls, name, bases, attrs)
+
+    @classmethod
+    def handle_grpc_errors(cls, func: Callable[..., Any]) -> Callable[..., Any]:
```

Comment:
Wondering if it makes more sense to have the DAL just raise gRPC errors, and have the error translations instead live on the superclass of Patient, Task, etc.

Reply:
Update: There is no need to worry about this now because this is encapsulated and can be changed later. The only thing that matters is the exception names.